sensor/common/heritage/heritage.go

```diff
@@ -154,63 +152,67 @@ func (h *Manager) SetCurrentSensorData(currentIP, currentContainerID string) {
+	h.currentSensor.UpdateTimestamps(now)
+
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 	defer cancel()
-	if err := h.upsertConfigMap(ctx, time.Now()); err != nil {
+	if _, err := h.upsertConfigMap(ctx, now); err != nil {
 		log.Warnf("Failed to update heritage data in the configMap: %v", err)
 	}
 }
 
-// updateCachedTimestampNoLock updates the timestamp if container ID and IP already exist in the heritage.
-// The size of h.cache is expected to be <10 in most of the cases.
-func (h *Manager) updateCachedTimestampNoLock(now time.Time) bool {
-	for _, entry := range h.cache {
+// getCurrentSensorIndexInCache checks whether an entry with given containerID and podIP exists in cache and returns its slice index.
+func (h *Manager) getCurrentSensorIndexInCache() (found bool, index int) {
+	for idx, entry := range h.cache {
 		if entry.ContainerID == h.currentSensor.ContainerID && entry.PodIP == h.currentSensor.PodIP {
-			if entry.SensorStart.IsZero() {
-				entry.SensorStart = now
-			}
-			entry.LatestUpdate = now
-			return true
+			return true, idx
 		}
 	}
-	return false
+	return false, 0
 }
 
-func (h *Manager) upsertConfigMap(ctx context.Context, now time.Time) error {
+func (h *Manager) upsertConfigMap(ctx context.Context, now time.Time) (wrote bool, err error) {
 	h.cacheMutex.Lock()
 	defer h.cacheMutex.Unlock()
 	if !h.cacheIsPopulated.Load() {
 		if err := h.populateCacheFromConfigMapNoLock(ctx); err != nil {
 			log.Warnf("%v", err)
 		}
 	}
-	wrote, err := h.write(ctx, now)
-	if wrote {
+	if updated := h.updateCache(now); !updated {
+		return false, nil
+	}
+	err = h.write(ctx)
+	if err == nil {
 		log.Debugf("Wrote heritage data %s to ConfigMap %s/%s", sensorMetadataString(h.cache), h.namespace, cmName)
 	}
-	return err
+	return true, err
 }
 
-func (h *Manager) write(ctx context.Context, now time.Time) (wrote bool, err error) {
-	if cacheHit := h.updateCachedTimestampNoLock(now); !cacheHit {
-		h.currentSensor.LatestUpdate = now
-		h.cache = append(h.cache, &h.currentSensor)
-	} else {
-		// On cache-hit, we would update the `LatestUpdate` field, which should not happen too often to save on IO load.
-		timeSinceLastWrite := now.Sub(h.lastCmWrite)
-		if timeSinceLastWrite < h.writeCmFrequency {
-			return false, nil
-		}
+// updateCache adds the currentSensor data to cache, prunes old data (according to settings) and returns true.
+// If the cache already contains data of currentSensor and there are no updates to its containerID or podIP,
+// then the configMap is not written and false is returned.
+func (h *Manager) updateCache(now time.Time) bool {
+	cacheHit, idx := h.getCurrentSensorIndexInCache()
+	if cacheHit {
+		h.cache[idx].UpdateTimestamps(now)
+		// Limit requests to k8s API and do not update configMap if only the timestamps have changed.
+		return false
 	}
 
+	// Current sensor is not found in cache or there is an update to containerID or podIP.
+	h.cache = append(h.cache, &h.currentSensor)
```

Comment:
Can we assume the cache is naturally sorted?

Reply:
Can you explain what you mean by that? What does it mean to naturally sort?